Hazel/Dtls/DtlsConnectionListener.cs

```diff
@@ -296,8 +296,14 @@ private void ProcessIncomingMessage(ByteSpan message, IPEndPoint peerAddress)
             PeerData peer = null;
             if (!this.existingPeers.TryGetValue(peerAddress, out peer))
             {
-                HandleNonPeerRecord(message, peerAddress);
-                return;
+                lock (this.existingPeers)
```

Comment:
This is the meat of the fix for bug 1.

Reply:
Great catch! This likely arose once we moved DTLS processing to multi-threads. Certainly adds to my desire to go lockless haha

Reply:
Yeah, this is one of the few places that I wonder if we'll ever be lockless. There's so much work that goes into creating the PeerData that it's usually not worth potentially wasting and then figuring out how to recover. This lock is parallel to one of the few locks left in the UDP layer as well. Would be interested in a proposal to remove it, though. I couldn't figure out how.

Reply:
Yup, I think we can still make it lockless! 

Basically we have to assign a mapping up from for IP to that thread's queue, so it would be a step before handling PeerData, so we can guarantee all PeerData handling is on the same thread (or maybe we can get away with checking if there's an existing connection readonly before queuing to the thread assigned to the peer). 

I do think there could be some refactors that might need to be placed on the UDP layer, particularly on the send side to be sure we are back on the same thread, but I think it is possible with an identifier